4C Integration Guide

edobry opened the pull request:
This document is meant to convey all the necessary information for 4C, or any other DSP (Demand-Side Platform), integrate and work with our API. Please read this document through that lens, and keep in mind that information about how the backend works internally, or certain intermediary concepts, are not relevant.

Please let me know if there's anything missing, incorrect, or extraneous. This is a first draft and is pending revision.

> For ease of reading, click this button on the "Files Changed" view:
![selection_131](https://user-images.githubusercontent.com/52695/46838771-c6d69c80-cd88-11e8-9210-b1ddfbf1d931.png)
You will need to switch back to the Source view to leave comments, however.

edobry committed:
added 4C doc

khabib9 commented:
Team,  Please make this review a priority. Please submit your comments by
12 noon tomorrow. Thanks a lot. H

On Thu, Oct 11, 2018, 7:08 PM Eugene Dobry <notifications@github.com> wrote:

> @edobry <https://github.com/edobry> requested your review on:
> massexchange/MXEngineering#5
> <https://github.com/massexchange/MXEngineering/pull/5> 4C Integration
> Guide.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/massexchange/MXEngineering/pull/5#event-1899533518>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ApmHgPgF8dbP6luFBnbhuobSGRzdvO6Wks5uj89UgaJpZM4XYhPt>
> .
>

ArtificerZeno commented on 4C Integration.md:
```diff
@@ -0,0 +1,262 @@
+
+* An `Organization` represents a single buyer or seller, and is analogous to a company account. There are many `Users` within an `Organization`, and any action a user takes looks from the outside as if it was done by the organization. Users are assigned to `Teams`, and given `Roles` within those teams which constrain which actions they are allowed to take.
+
+* A `Market` is an environment in which transactions happen. There is an `Open Market` and several `Private Markets`. Initially, all transactions will happen in the latter. `Orders` in private markets can only be seen by market members, which are `Organizations`, and defined by the market owner.
```
maybe say 'user made' instead of 'several'. 'several' seems finite and it will help with 'defined by market owner' later in the paragraph

ArtificerZeno commented on 4C Integration.md:
```diff
@@ -0,0 +1,262 @@
+
+* An `Order` is how we represent supply or demand entered into a `Market`. An order represents the intention of an `Organization` to buy/sell some amount of a certain `Asset` in a `Market`, at some price.
+
+* An `Asset` is what we call media that is bought and sold. It is composed of `Attributes` and additional information, like the date and time. An `Attribute` represents some piece of information about the asset. `Attributes` in turn have `Types`, which can be thought of as classifications of information about assets. This is displayed in the UI in the format of `Type: Attribute Value`, ie, `Netwo ...
```
maybe this paragraph before the one above. Define asset before it is used in order?

edobry replied:
i was considering that, but settled on this order, as i felt it was important to define `Order` right after `Market`, and that inserting `Asset` in between would be a non-sequitor. I figured that since `Asset` is defined immediately after the first time its used, it would be fine. what do you think?

ArtificerZeno replied:
that's a good point, agreed.

ArtificerZeno commented on 4C Integration.md:
```diff
@@ -0,0 +1,262 @@
+
+* The market operates on **New York time (EDT, GMT-5)**, and is open from **9:30AM-4:30PM**. Orders may be submitted at any time, but will **only be processed** when the market is open.
+
+* Orders are processed in a **first-in-first-out** queue; if your order is submitted before another one, it is guaranteed to get access to supply first. If an order is edited, its position in the queue is reset.
```
maybe 'If an order is edited, it becomes last in queue' instead? 'reset' seems a little ambiguous

RobertTUnnold reviewed (approved):
Looks ok to me

ArtificerZeno commented on 4C Integration.md:
```diff
@@ -0,0 +1,262 @@
+* flightStartDate: `Date`
+* flightEndDate: `Date`
+
+##### `OrderGroup`
```
is it safe to ignore status here?

edobry replied:
resolved

ArtificerZeno commented on 4C Integration.md:
```diff
@@ -0,0 +1,262 @@
+
+Request body: `OrderGroup`
+
+##### Buy Order Status Query
```
status is a little misleading here? maybe 'Buy Order Query' and the next line is 'Query for the current version of your OrderGroup'?

ArtificerZeno reviewed (approved):
I have some small things (oddly 3 of them are language things :v) but I don't see them as being important enough to stop this from going through

aulman committed:
Only minor punctuation corrections

revenueStream: `RevenueStream`  and * rateCardDollars: `Currency` in ##### `Deal` are AMC Network internal fields and should not be exposed. I reached out to @Karl to and he verified.

aulman committed:
Update 4C Integration.md

aulman reviewed (approved):
revenueStream: `RevenueStream`  and * rateCardDollars: `Currency` in ##### `Deal` are AMC Network internal fields and should not be exposed. I reached out to @karl to and he verified. Removed.

Additional minor punctuation corrections.

edobry committed:
addressed order group concerns and typos

edobry committed:
added more domain information

edobry committed:
removed components section

edobry committed:
improved market rules and order semantics

edobry committed:
removed all Log references

edobry committed:
added an example

edobry commented:
I've added on example and am working on adding more. Please take a look and let me know what you think, if there's anything I should add or if it is unclear.

edobry committed:
removed references to AMC

edobry committed:
added campaign example

edobry committed:
renamed OrderGroup to BuyOrder, added market endpoint

edobry committed:
added more examples

edobry committed:
fixed github newline handling

edobry commented:
I've added more examples and revised some parts; pls review

edobry committed:
clarified order priority, partial fill, and expiration

edobry committed:
added order statuses

edobry committed:
added buy order deactivation endpoint and example

edobry committed:
added info on all or none orders

edobry committed:
clarified order cancellation timing and partial match rules

ArtificerZeno commented on 4C Integration.md:
```diff
@@ -466,3 +478,17 @@ Take the `marketId` and `groupId` from the previous step, substitute, and then m
+
+Sometimes, you'll want to deactivate an order you've submitted, either if you made a mistake, changed your mind, or have filled your demand elsewhere.
+
+The `BuyOrder` deactivation endpoint is `/orderGroup/{groupId}/status` ad it accepts `PUT` requests with an `OrderStatus` in the request body. The only currently accepted status is `Deactivated`.
```
'ad it accepts' seems to be a typo

edobry committed:
added match summary and log/adv endpoints

edobry committed:
added match summary query example

edobry committed:
Revert "removed all Log references"

This reverts commit 54ba27a978ae30d4d05d4f782f92b7462912e17b.

# Conflicts:
#	4C Integration.md

edobry committed:
added log example

edobry committed:
typo
